### PR TITLE
MULE-9985 Remove NullPayload

### PR DESCRIPTION
**Pull request by dfeist**

### CODE REVIEW

File: core-tests/src/test/java/org/mule/runtime/core/MuleMessageDataTypePropagationTestCase.java [hunk 1]
```diff
@@ -109,15 +107,15 @@ public void updatesDataTypeWhenPayloadIsReplacedWithNullPayload() throws Excepti
     {
         MuleMessage muleMessage = MuleMessage.builder().payload(TEST).build();
 
-        assertDataType(MuleMessage.builder(muleMessage).payload(NullPayload.getInstance()).build(), Object.class, ANY, null);
+        assertDataType(MuleMessage.builder().nullPayload().build(), Object.class, ANY, null);
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/MuleMessageDataTypePropagationTestCase.java]: why was the cloning removed? if it's not needed anymore, remove the previous line too.

File: core-tests/src/test/java/org/mule/runtime/core/message/DefaultMuleMessageBuilderTestCase.java [hunk 2]
```diff
@@ -354,21 +349,17 @@ public void outboundAttachment()
...
-    @Rule
-    public ExpectedException thrown = ExpectedException.none();
-
     @Test
-    public void nullPayload()
+    public void nullPayloadDoesNotFail()
     {
         DefaultMuleMessageBuilder builder = new DefaultMuleMessageBuilder();
-        thrown.expect(NullPointerException.class);
         builder.payload(null);
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/message/DefaultMuleMessageBuilderTestCase.java]: Is this still valid? Shouldn't it be a call to nullPayload()?

File: core-tests/src/test/java/org/mule/runtime/core/mule/model/MethodHeaderEntryPointResolverTestCase.java [hunk 3]
```diff
@@ -48,7 +46,7 @@ public void testMethodSetPass() throws Exception
     @Test
     public void testMethodSetWithNoArgsPass() throws Exception
     {
-        MuleEventContext ctx = createMuleEventContext(getInstance(), singletonMap("method", "wash"));
+        MuleEventContext ctx = createMuleEventContext(null, singletonMap("method", "wash"));
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/mule/model/MethodHeaderEntryPointResolverTestCase.java]: why this change?

File: core-tests/src/test/java/org/mule/runtime/core/MuleMessageDataTypePropagationTestCase.java (hunk 1 above)
Comment (dfeist) [inline core-tests/src/test/java/org/mule/runtime/core/MuleMessageDataTypePropagationTestCase.java]: mistake. fixing..

File: core-tests/src/test/java/org/mule/runtime/core/transformer/AbstractTransformerTestCase.java [hunk 4]
```diff
@@ -54,10 +54,8 @@ public void testTransform() throws Exception
     {
         Transformer trans = this.getTransformer();
         Object result = trans.transform(getTestData());
-        assertNotNull("The result of the transform shouldn't be null", result);
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/transformer/AbstractTransformerTestCase.java]: why are these assertions removed?

File: core/src/main/java/org/mule/runtime/core/api/MessagingException.java [hunk 5]
```diff
@@ -149,33 +148,38 @@ protected String generateMessage(Message message, MuleContext muleContext)
...
-                    addInfo(PAYLOAD_TYPE_INFO_KEY, muleMessage.getPayload().getClass().getName());
-                    try
+                    if (payload != null)
                     {
-                        addInfo(PAYLOAD_INFO_KEY, muleContext.getTransformationService().transform(muleMessage, DataType.STRING).getPayload());
+                        addInfo(PAYLOAD_TYPE_INFO_KEY, muleMessage.getDataType().getType().getName());
+                        try
+                        {
+                            addInfo(PAYLOAD_INFO_KEY, muleContext.getTransformationService().transform(muleMessage, DataType.STRING).getPayload());
+
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/MessagingException.java]: extra line

File: core/src/main/java/org/mule/runtime/core/api/context/notification/ServerNotification.java [hunk 6]
```diff
@@ -86,7 +87,7 @@ public ServerNotification(Object message, int action)
 
     public ServerNotification(Object message, int action, String resourceIdentifier)
     {
-        super((message == null ? NULL_MESSAGE : message));
+        super((message == null ? MuleMessage.builder().nullPayload().build() : message));
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/context/notification/ServerNotification.java]: should the NULL_MESSAGE constant be removed too?

File: core/src/main/java/org/mule/runtime/core/component/AbstractComponent.java [hunk 7]
```diff
@@ -165,7 +164,7 @@ else if (result instanceof VoidResult)
         {
             return event;
         }
-        else if (result != null)
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/component/AbstractComponent.java]: can't this happen anymore?

File: core/src/main/java/org/mule/runtime/core/expression/transformers/ExpressionTransformer.java [hunk 8]
```diff
@@ -80,7 +79,7 @@ private boolean checkIfAllAreNull(Object[] objects)
     {
         for (int i = 0; i < objects.length; i++)
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/expression/transformers/ExpressionTransformer.java]: change to enhanced for

File: transports/jms/src/test/java/org/mule/compatibility/transport/jms/JmsTransactionalCachingTestCase.java [hunk 9]
```diff
@@ -46,7 +44,7 @@ public void cachesSession() throws Exception
         MuleMessage response = client.send("vm://testInput", TEST_MESSAGE_1, null);
         assertThat(TEST_MESSAGE_1, equalTo(getPayloadAsString(response)));
         response = client.send("vm://testInput", TEST_MESSAGE_2, null);
-        assertThat(NullPayload.getInstance(), equalTo(response.getPayload()));
+        assertThat(null, equalTo(response.getPayload()));
```
Comment (elrodro83) [inline transports/jms/src/test/java/org/mule/compatibility/transport/jms/JmsTransactionalCachingTestCase.java]: invert this assertion
